pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java

```diff
@@ -943,17 +943,20 @@ void receiveIndividualMessagesFromBatch(MessageMetadata msgMetadata, int redeliv
                 if (subscriptionMode == SubscriptionMode.NonDurable && startMessageId != null
                         && messageId.getLedgerId() == startMessageId.getLedgerId()
                         && messageId.getEntryId() == startMessageId.getEntryId()
-                        && i <= startMessageId.getBatchIndex()) {
+                        && i < startMessageId.getBatchIndex()) {
                     // If we are receiving a batch message, we need to discard messages that were prior
                     // to the startMessageId
                     if (log.isDebugEnabled()) {
-                        log.debug("[{}] [{}] Ignoring message from before the startMessageId", subscription,
-                                consumerName);
+                        log.debug("[{}] [{}] Ignoring message from before the startMessageId: {}", subscription,
+                                consumerName, startMessageId);
                     }
+                    singleMessagePayload.release();
```

Comment:
Good catch

Reply:
@lovelle Can you commit this bug fix on a separate PR? So that we can the new feature on its own.

Reply:
Yep, sure.

Reply:
Done: https://github.com/apache/pulsar/pull/4342